<launch>

  <!-- CMax-SLAM node -->
  <node name="cmax_slam" pkg="cmax_slam" type="cmax_slam" output="screen" args="--v 0">

    <!-- Topics -->
    <param name="events_topic" value="/dvs/events" />
    <param name="camera_info_topic" value="/dvs/camera_info" />

    <!-- COMMON PARAMS -->
    <!-- Objective function to be optimized -->
    <param name="contrast_measure" value="0" /> <!-- 0=Variance, 1=Mean Square, etc.-->
    <!-- Event warping options -->
    <param name="frontend_blur_sigma" value="1.0" /> <!-- Gaussian blurring for the local CMax, in front-end -->
    <param name="backend_blur_sigma" value="1.0" /> <!-- Gaussian blurring for the global CMax, in back-end -->
    <param name="event_batch_size" value="100" /> <!-- Number of events in each event mini-batch -->
    <!-- Event subsampling -->
    <param name="frontend_event_sample_rate" value="10" /> <!-- Process 10% of all events in the front-end -->
    <param name="backend_event_sample_rate" value="5" /> <!-- Process 2% of all events in the back-end -->

    <!-- FRONT-END PARAMS -->
    <param name="num_events_per_packet" value="5000" /> <!-- Depends on the camera resolution, motion speed, and event sampling rate -->
    <param name="dt_ang_vel" value="0.04" /> <!-- Angular velocity frequency f = 1/dt_ang_vel -->
    <param name="show_local_iwe" value="true" /> <!-- Whether to publish local IWEs -->

    <!-- BACK-END PARAMS -->
    <!-- Sliding window settings-->
    <param name="backend_time_window_size" value="0.1" /> <!-- Size of the sliding window [second] -->
    <param name="backend_sliding_window_stride" value="0.1" /> <!-- Stride of the sliding window [second] -->
    <!-- Trajectory settings -->
    <param name="spline_degree" value="1" /> <!-- 1: Linear, 3: Cubic -->
    <param name="dt_knots" value="0.1" />
    <!-- Pano map settings -->
    <param name="pano_height" value="256" /> <!-- The bigger the map, the slower the speed of CMax-SLAM -->
    <param name="backend_min_ev_rate" value="10000" /> <!-- Minimal event rate to launch the backend [ev/s] -->
    <param name="max_update_times" value="100" /> <!-- Max updating times for the panoramic map (updating frequency: 20Hz -->
    <param name="Y_angle" value="0.0" /> <!-- Set the starting point of the camera FOV (initial yaw) [deg] -->
    <param name="gamma" value="0.75" /> <!-- Gamma correction coefficient, for bettervisualization -->
    <!-- Visualization -->
    <param name="show_pano_map" value="true" /> <!-- Whether to publish global panoramic map -->
    <param name="draw_FOV" value="true" /> <!-- Whether to draw the camera FOV on the panoramic map -->
  </node>

  <!-- Display the panoramic map using rqt_image_view -->
  <node name="pano_map" pkg="rqt_image_view" type="rqt_image_view" />

</launch>
